### PR TITLE
8293566: RISC-V: Clean up push and pop registers

### PR DESCRIPTION
**Pull request by feilongjiang**
Currently, riscv port has two implementations of `push_reg` and `pop_reg`.
The bitset version can be replaced with the RegSet version for better readability.

For example, if we want to push x10 and x11, pushing register with bitset would be like:
```
push_reg(0xc00, sp)
```

while pushing register with RegSet will be more clear:
```
push_reg(RegSet::of(x10, x11), sp);
```

Testing:

- hotspot and jdk ter1 on QEMU without new failures

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8293566](https://bugs.openjdk.org/browse/JDK-8293566): RISC-V: Clean up push and pop registers

### Reviewers
 * [Fei Yang](https://openjdk.org/census#fyang) (@RealFYang - **Reviewer**)
 * [Aleksey Shipilev](https://openjdk.org/census#shade) (@shipilev - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10227/head:pull/10227` \
`$ git checkout pull/10227`

Update a local copy of the PR: \
`$ git checkout pull/10227` \
`$ git pull https://git.openjdk.org/jdk pull/10227/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10227`

View PR using the GUI difftool: \
`$ git pr show -t 10227`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10227.diff">https://git.openjdk.org/jdk/pull/10227.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/cpu/riscv/macroAssembler_riscv.cpp [hunk 1]
```diff
@@ -946,59 +946,53 @@ int MacroAssembler::bitset_to_regs(unsigned int bitset, unsigned char* regs) {
...
-// Return the number of words pushed
-int MacroAssembler::push_reg(unsigned int bitset, Register stack) {
+// Push lots of registers in the regset supplied.  Don't push sp.
+void MacroAssembler::push_reg(RegSet regset, Register stack) {
   DEBUG_ONLY(int words_pushed = 0;)
   CompressibleRegion cr(this);
 
   unsigned char regs[32];
-  int count = bitset_to_regs(bitset, regs);
+  int count = bitset_to_regs(regset.bits(), regs);
```
Comment (RealFYang) [inline src/hotspot/cpu/riscv/macroAssembler_riscv.cpp:955]: I think it's better to have a zero-check for regs.bits() before we run into the loop in MacroAssembler::bitset_to_regs. So I would suggest make the old API "int MacroAssembler::push_reg(unsigned int bitset, Register stack)" a private member function and implement the new API as:

void MacroAssembler::push_reg(RegSet regset, Register stack) { if (regs.bits()) push_reg(regs.bits(), stack); }

Similar for pop_reg and other APIs here.

File: src/hotspot/cpu/riscv/macroAssembler_riscv.hpp [hunk 2]
```diff
@@ -473,17 +473,15 @@ class MacroAssembler: public Assembler {
...
-  void pop_fp(FloatRegSet regs, Register stack) { if (regs.bits()) pop_fp(regs.bits(), stack); }
+  void push_reg(RegSet regset, Register stack);
+  void pop_reg(RegSet regset, Register stack);
+  void push_fp(FloatRegSet regset, Register stack);
+  void pop_fp(FloatRegSet regset, Register stack);
 #ifdef COMPILER2
-  void push_vp(VectorRegSet regs, Register stack) { if (regs.bits()) push_vp(regs.bits(), stack); }
-  void pop_vp(VectorRegSet regs, Register stack) { if (regs.bits()) pop_vp(regs.bits(), stack); }
+  void push_vp(VectorRegSet regset, Register stack);
+  void pop_vp(VectorRegSet regset, Register stack);
```
Comment (RealFYang) [inline src/hotspot/cpu/riscv/macroAssembler_riscv.hpp:484]: The function name push_vp / pop_vp look a bit strange to me. I think we should rename them to push_v / pop_v.

**Review by RealFYang — CHANGES_REQUESTED** [review]
Thanks for the cleanup. I have two comments here.

File: src/hotspot/cpu/riscv/macroAssembler_riscv.cpp (hunk 1 above)
Comment (feilongjiang) [reply src/hotspot/cpu/riscv/macroAssembler_riscv.cpp]: Done, keep the old bitset verison and make them private.

File: src/hotspot/cpu/riscv/macroAssembler_riscv.hpp (hunk 2 above)
Comment (feilongjiang) [reply src/hotspot/cpu/riscv/macroAssembler_riscv.hpp]: Done, renamed to `push_v/pop_v`.

**Review by RealFYang — APPROVED** [review]
Updated change looks good. Thanks.

**Review by shipilev — APPROVED** [review]
Looks good to me.

**Comment by feilongjiang** [comment]
@RealFYang @shipilev --Thanks for reviewing!
/integrate

**Pull request closed without merging**